### PR TITLE
SearchControl: refactor to use InputControl internally

### PR DESCRIPTION
**Pull request by ciampo**
Part of #56388

## What?

Changes the implementation of the SearchControl so that it uses InputControl internally.

The size variants, which were 48px (default), 40px (next default), and 32px (compact); are now just 40px (default) and 32px (compact).

## Why?

To improve cohesion across the app, as discussed in #56388.

## Testing Instructions

See Storybook stories for SearchControl.

Please review consumer changes in #58014, which is stacked onto this PR.

## Screenshots or screencast 

| Before | After |
|--------|-------|
|Old search control|New search control|

### CODE REVIEW

**Comment by jameskoster** [comment]
Hey @ciampo do you have any plans to pick this back up? It would be very handy for the data views work. Let me know if I can help at all.

**Comment by ciampo** [comment]
Apologies for not making much progress here, I've been busy with other tasks. I'll try to look into it soon (or ask for help otherwise)!

File: packages/components/src/search-control/index.tsx
Comment (mirka) [inline packages/components/src/search-control/index.tsx:55]: I noticed that this component didn't have a default label, and since it was not marked as required in TS and visually hidden by default, there were likely a lot of unlabeled usages 😔 I gave it a default and updated the docs.

File: packages/components/src/search-control/types.ts
Comment (mirka) [inline packages/components/src/search-control/types.ts:11]: I want to take this opportunity to make this the default now. It used to be 48px by default, which is completely outside our sizing scheme and larger than the new default, so I don't anticipate this to be a problem.

File: packages/components/src/search-control/types.ts
Comment (mirka) [inline packages/components/src/search-control/types.ts:50]: We need to specifically inherit these three event handler types from InputControl because they are different from standard `<input>` event handlers. However, they are irrelevant to SearchControl so we should hide them from the docs.

Comment (ciampo) [reply packages/components/src/search-control/types.ts:50]: TIL the `@ignore` tag

**Comment by jameskoster** [comment]
In terms of design this is looking pretty good. I think the only adjustment I'd suggest is to reduce the inner horizontal padding on the input slightly. At the current distance (8px) the icon is a bit too far from the text. 4px should work.

File: packages/components/src/search-control/index.tsx
Comment (ciampo) [inline packages/components/src/search-control/index.tsx]: I had somehow forgot about using the `_overrides` key 🤯

File: packages/components/src/search-control/index.tsx
Comment (ciampo) [inline packages/components/src/search-control/index.tsx]: Currently, `Suffixitem` is being re-created every time `SearchControl` re-renders. Should we move it out of the render function and pass it the necessary props (`onClose`, `value`, `size`, `onChange`, `searchRef` ) ? Or alternatively wrap it in `React.memo()` ?

File: packages/components/src/search-control/types.ts
Comment (ciampo) [reply packages/components/src/search-control/types.ts:11]: I guess that's ok, given that the previous height was different from other control components anyway.

**Review by ciampo — COMMENTED** [review]
Code changes overall look good.

The fact that we don't have dedicated unit test means that we'll have to do some extra manual testing to make sure that no regressions are introduced.

It would be also great to get feedback from @WordPress/gutenberg-design around the new look of the component.

File: packages/components/src/search-control/index.tsx
Comment (mirka) [reply packages/components/src/search-control/index.tsx]: Extracted to pure function in 93a2b4c

**Comment by mirka** [comment]
> I think the only adjustment I'd suggest is to reduce the inner horizontal padding on the input slightly. At the current distance (8px) the icon is a bit too far from the text. 4px should work.

@jameskoster Done 👍 The screenshot in the PR description is updated with the new spacing.

@ciampo Feedback is addressed now 👍 I added unit tests in #58693 and confirmed they pass on this branch.

Please proceed to review #58014, which will be required to land this PR (it will be merged into this branch first before landing). Both branches are now up-to-date with current trunk as well (rebased).

**Comment by ciampo** [comment]
#58693 and #58014 reviewed and approved.

I guess the next steps could be:
- to merge both PRs
- to rebase on top of `trunk` and enable controlled unit tests for the component

We should then be in a good place to approve and merge this PR too!

**Comment by jasmussen** [comment]
The gray background was an intentional design choice as part of the inserter work, so let's not get rid of that. It can exist as an option if need be, but for now let's not redesign this, just align the functionality.

**Comment by mirka** [comment]
After discussing with the design team, we won't be doing any design changes at this time (9c18f1e). I've made some adjustments so the "borderless gray" input can be achieved in a sustainable way (e91440cd3b2cc1d83fc56a8cfb2d428884c92888).

On hold until #58750 lands, so we can do the borderless here.

**Comment by ciampo** [comment]
Sounds good. Will test again once changes from #58750 are integrated

File: packages/components/src/search-control/index.tsx
Comment (mirka) [inline packages/components/src/search-control/index.tsx]: I added this little safe guard so consumers don't improvise a `disabled` variant of the SearchControl. Currently, the light gray background of the search box is exactly the same color as a disabled InputControl, so we're going to need some coordinated design work if anyone ever needs a disabled search box.

**Comment by mirka** [comment]
Everything is now up to date and working. The screenshot in the PR description is also updated 👍

File: packages/components/src/search-control/index.tsx
Comment (tyxla) [inline packages/components/src/search-control/index.tsx:105]: Nit: sometimes we can see `value || ''` and sometimes `value ?? ''`. While for a normal input control, there shouldn't be a big difference, perhaps we can make these consistent.

File: packages/components/src/search-control/index.tsx
Comment (tyxla) [inline packages/components/src/search-control/index.tsx]: Maybe we should mention the team handle? `@WordPress/gutenberg-design`?

**Review by tyxla — APPROVED** [review]
Tests well and looks good 👍 

I think it's worth adding as an `Internal` entry in the CHANGELOG.

🚀

File: packages/components/src/search-control/index.tsx
Comment (mirka) [reply packages/components/src/search-control/index.tsx:105]: Good idea, I'll unify to `??` since they're closer to the intent. The only value that it would affect here is an empty string, and since we're falling back to an empty string it shouldn't change the outcome. 62b49c0e

File: packages/components/src/search-control/index.tsx
Comment (mirka) [reply packages/components/src/search-control/index.tsx]: Nice 👍 5247c854

**Comment by youknowriad** [comment]
While I can't say for sure 100%, I think this PR impacted the "inserter open" performance metric negatively (I'd say around 6 or 7%). Would be cool to try to debug a bit.

**Comment by youknowriad** [comment]
Inserter search is impacted as well. 

For clarity, we have a gap of 24h where the performance tests were not working, so we can't pin point exactly which commit introduced the regression by looking at the graphs https://www.codevitals.run/project/gutenberg 

I tried looking at the commits and this one is the most suspicious for me (looking at the PR performance job, it seems to confirm, although a single performance job run is not sufficient to judge entirely)

**Comment by mirka** [comment]
A little bit of performance hit is expected, since we replaced a native `input` with an `InputControl`, which simply processes more JS code. It looks like we're adding 1–2 ms to the inserter metrics, which if I were to guess is mostly the difference between `input` and `InputControl`.

**Comment by youknowriad** [comment]
For me, That percentage in the increase is too much for a single component added to the react tree. If we had that performance loss everytime we add a new component to the react tree of the inserter, I think we'll be in a very bad position. Just imagine the number of components we have in the inserter. If one of them is responsible of 7% (something like 12ms or something like that in the graph) that seems like a component that we should investigate

**Comment by tyxla** [comment]
Agreed that we should be mindful of performance. I wonder if this is something we can pinpoint, be it in `InputControl` or the underlying components. Could be a good idea to do some measurements in isolation and see if something will stand out.

**Comment by mirka** [comment]
I've been doing some isolated perf testing on the components, benchmarking InputControl against TextControl (which is significantly simpler and about 30–40% more performant). Initial experimentation seems to show that the main drag is Emotion, or at least the specific way we use it to add dynamic prop-based styles. I'll see what I can do.

**Comment by mirka** [comment]
I summarized my findings and next steps [here](https://github.com/WordPress/gutenberg/pull/58948#issuecomment-1942040476).

When we do start migrating, we can start with InputControl since it is widely used, and hopefully we'll be able to validate the gains through Code Vitals.

**Comment by youknowriad** [comment]
Thanks for diving into this @mirka I appreciate it.

**Pull request merged**
